### PR TITLE
Issue #4191 - GzipHandler.minGzipSize cannot be lower then 23.

### PR DESCRIPTION
**Pull request by joakime**
#4191
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 1]
```diff
@@ -156,7 +155,7 @@
 {
     public static final String GZIP = "gzip";
     public static final String DEFLATE = "deflate";
-    public static final int DEFAULT_MIN_GZIP_SIZE = 16;
+    public static final int DEFAULT_MIN_GZIP_SIZE = 23;
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: While 23 is the size break even point, the CPU break even point is probably a lot higher.  I'd be OK with us setting this to be something like 1024, which is close to a MTU, so sending a single TCP frame can often be about the same cost regardless of size.

**Comment by gregw** [comment]
Since @joakime is away today, I made the changes that I would like to see.   @janbartel can you review?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 2]
```diff
@@ -94,7 +93,7 @@
  * <li>
  * Is the Response {@code Content-Length} header present, and does its
  * value meet the minimum gzip size requirements?
- * <br> (Default: 16 bytes. see {@link GzipHandler#DEFAULT_MIN_GZIP_SIZE})
+ * <br> (Default: 1024 bytes. see {@link GzipHandler#BREAK_EVEN_GZIP_SIZE})
```
Comment (lachlan-roberts) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: It seems like `DEFAULT_MIN_GZIP_SIZE` and `BREAK_EVEN_GZIP_SIZE` are independent of each other and you should have both. This javadoc doesn't make as much sense as its not linking to the default min gzip size here.

**Comment by lachlan-roberts** [comment]
@gregw `IncludedGzipTest` is failing.

**Comment by gregw** [comment]
@lachlan-roberts should be fixed now

**Comment by joakime** [comment]
> Since @joakime is away today, I made the changes that I would like to see. @janbartel can you review?

I'm not away today.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 3]
```diff
@@ -156,7 +154,7 @@
 {
     public static final String GZIP = "gzip";
     public static final String DEFLATE = "deflate";
-    public static final int DEFAULT_MIN_GZIP_SIZE = 16;
+    public static final int BREAK_EVEN_GZIP_SIZE = 23;
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: This is a public API, don't change this name.
Or at least don't delete the original name.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 4]
```diff
@@ -165,7 +163,7 @@
     private int _poolCapacity = -1;
     private DeflaterPool _deflaterPool = null;
 
-    private int _minGzipSize = DEFAULT_MIN_GZIP_SIZE;
+    private int _minGzipSize = 1024;
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: This should be 32, not 1024.

Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: There's TONs of content in the REST space that lives above 32 and below 1024.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 5]
```diff
@@ -946,13 +944,19 @@ public void setIncludedPaths(String... pathspecs)
...
+     * </p>
      *
-     * @param minGzipSize minimum response size in bytes
+     * @param minGzipSize minimum response size in bytes (not allowed to be lower then 23)
      */
     public void setMinGzipSize(int minGzipSize)
     {
-        _minGzipSize = minGzipSize;
+        if (minGzipSize < BREAK_EVEN_GZIP_SIZE)
+            LOG.warn("minGzipSize {} is less break even size {}", minGzipSize, BREAK_EVEN_GZIP_SIZE);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: `("minGzipSize of {} is inefficient for short content, break even is size {}", minGzipSize, BREAK_EVEN_GZIP_SIZE)`
or
`("minGzipSize of {} is less than break even size of {}, which is inefficient for short content", minGzipSize, BREAK_EVEN_GZIP_SIZE)` 
?

File: jetty-servlets/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipDefaultTest.java [hunk 6]
```diff
@@ -468,6 +468,7 @@ public void testGzipCompressedByContentTypeWithEncoding() throws Exception
         tester.setContentServlet(HttpContentTypeWithEncoding.class);
         tester.getGzipHandler().addIncludedMimeTypes("text/plain");
         tester.getGzipHandler().setExcludedAgentPatterns();
+        tester.getGzipHandler().setMinGzipSize(16);
```
Comment (joakime) [inline jetty-servlets/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipDefaultTest.java]: I bet if you had a testcase in this actually producing 17 bytes of input content you'd see the break even issue.

**Comment by gregw** [comment]
@joakime Oh - then review my change :)     merge if you're happy... if not let's chat at tomorrows meeting.

File: jetty-servlets/src/test/java/org/eclipse/jetty/servlets/GzipFilterLayeredTest.java [hunk 7]
```diff
@@ -61,7 +61,7 @@
     private static final HttpConfiguration defaultHttp = new HttpConfiguration();
     private static final int LARGE = defaultHttp.getOutputBufferSize() * 8;
     private static final int SMALL = defaultHttp.getOutputBufferSize() / 4;
-    private static final int TINY = GzipHandler.DEFAULT_MIN_GZIP_SIZE / 2;
+    private static final int TINY = GzipHandler.BREAK_EVEN_GZIP_SIZE / 2;
```
Comment (joakime) [inline jetty-servlets/src/test/java/org/eclipse/jetty/servlets/GzipFilterLayeredTest.java]: Again, DEFAULT_MIN_GZIP_SIZE is a public API, don't change it for this version of Jetty.

File: jetty-servlets/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipContentLengthTest.java [hunk 8]
```diff
@@ -54,7 +54,7 @@
     private static final int LARGE = defaultHttp.getOutputBufferSize() * 8;
     private static final int MEDIUM = defaultHttp.getOutputBufferSize();
     private static final int SMALL = defaultHttp.getOutputBufferSize() / 4;
-    private static final int TINY = GzipHandler.DEFAULT_MIN_GZIP_SIZE / 2;
+    private static final int TINY = GzipHandler.BREAK_EVEN_GZIP_SIZE / 2;
```
Comment (joakime) [inline jetty-servlets/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipContentLengthTest.java]: DEFAULT_MIN_GZIP_SIZE is a public API, don't change it for this version of Jetty.

**Comment by gregw** [comment]
@joakime fixes from your review pushed.

**Pull request merged**